unittests/Basic/DiagnosticTest.cpp

```diff
@@ -71,6 +71,34 @@ TEST(DiagnosticTest, fatalsAsError) {
     EXPECT_EQ(Diags.getNumWarnings(), FatalsAsError);
   }
 }
+// Check that soft RESET works as intended
+TEST(DiagnosticTest, softReset) {
+  DiagnosticsEngine Diags(new DiagnosticIDs(),
+                          new DiagnosticOptions,
+                          new IgnoringDiagConsumer());
+
+  unsigned numWarnings = 0U, numErrors = 0U, delayedDiagID = 0U;
+
+  Diags.Reset(true);
+  // Check For ErrorOccurred and TrapNumErrorsOccurred
+  EXPECT_FALSE(Diags.hasErrorOccurred());
+  EXPECT_FALSE(Diags.hasFatalErrorOccurred());
+  EXPECT_FALSE(Diags.hasUncompilableErrorOccurred());
+  // Check for UnrecoverableErrorOccurred and TrapNumUnrecoverableErrorsOccurred
+  EXPECT_FALSE(Diags.hasUnrecoverableErrorOccurred());
+
+  EXPECT_EQ(Diags.getNumWarnings(), numWarnings);
+  EXPECT_EQ(Diags.getNumErrors(), numErrors);
+  EXPECT_EQ(Diags.getDelayedDiagID(), delayedDiagID);
+
+  EXPECT_FALSE(Diags.isDiagnosticInFlight());
+  EXPECT_TRUE(Diags.isLastDiagnosticIgnored());
+
+  EXPECT_FALSE(Diags.getDiagStatesEmpty());
+  EXPECT_FALSE(Diags.getDiagStatesByLocEmpty());
+  EXPECT_FALSE(Diags.getDiagStateOnPushStack());
```

Comment:
I feel these getters break encapsulation and the testcase maybe not a strong enough argument for them. Can we indirectly compare the state by using `DiagnosticsEngine::GetCurDiagState` before and after the reset?

Reply:
Only a check on `DiagStatesByLoc` can re-use some existing LOCs. Yes, we can have a call like `DiagStatesByLoc.GetCurDiagState()`, in the new method `getDiagStatesByLocEmpty` it will [return](https://code.woboq.org/llvm/clang/include/clang/Basic/Diagnostic.h.html#_ZNK5clang17DiagnosticsEngine12DiagStateMap15getCurDiagStateEv) `CurDiagState` -> check for `NULL`.

Everything else, at the core, that distinguishes `soft` reset, is !`private`!. 

OR

We can have a public variable just for `soft` Reset success. Set it with these in `if()` and create an accessible `getter`. (waste/no use otherwise).

OR 

There cannot be a `soft` `Reset` Test or `Reset` success test.

Reply:
I find https://google.github.io/googletest/reference/testing.html#FRIEND_TEST which might do what we need...

Reply:
ok, i'll [see](https://github.com/llvm/llvm-project/search?q=FRIEND_TEST&type=) and add.

Reply:
None of the tests are present in `namespace clang`. I don't think we can opt for https://google.github.io/googletest/reference/testing.html#FRIEND_TEST.

Reply:
I think we could move the testcase in that namespace, but if we cannot we are left with `#define private public` before inclusion of the header.